### PR TITLE
Update Mesa to 20xx, switch build system to use Meson

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

**Comment by jdm** [comment]
You may want to steal from https://github.com/servo/osmesa-src/pull/40/commits/a3e7ddc0e3b192a02d8e46232311bc4519f002e4 where I previously got CI passing.

**Comment by kvark** [comment]
@jdm wrong commit?

File: build.rs [hunk 1]
```diff
@@ -1,73 +1,37 @@
...
+        .arg("-Dgbm=disabled")
+        .arg("-Dglx=disabled"));
 
-    run(Command::new(find_make())
-                .env("MAKEFLAGS", env::var("CARGO_MAKEFLAGS").unwrap_or_default())
-                .env("PYTHONPATH", src.join("Mako-1.0.7.zip"))
-                .current_dir(&dst));
+    run(Command::new("ninja")
+        .current_dir(&build_folder)
+        );
```
Comment (emilio) [inline build.rs]: Formatting here looks off.

**Review by emilio — COMMENTED** [review]
Could you update `./tools/update-mesa.sh` with the process you use to update the repo?

Also, from what I can see, we can probably get away with running meson at import time (like `./tools/update-mesa.sh` does, and then require just `ninja`)

**Comment by jdm** [comment]
Indeed, I meant https://github.com/servo/osmesa-src/pull/37/commits/96d92bd67cc764fcaebc803e8ca5e5018f59ec03.

**Comment by kvark** [comment]
> Could you update ./tools/update-mesa.sh with the process you use to update the repo?

it is updated already

**Comment by emilio** [comment]
Ah, it seems it was hidden from the GitHub review thing. I'll pull instead :)

File: Cargo.toml [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
 authors = ["mesa3d maintainers"]
 build = "build.rs"
 repository = "https://github.com/servo/osmesa-src/"
 license = "MIT"
 description = "Mesa is a 3-D graphics library with an API which is very similar to that of OpenGL."
 homepage = "http://mesa3d.org"
+edition = "2018"
+
+[build-dependencies]
+meson = "1"
```
Comment (emilio) [inline Cargo.toml]: This crate doesn't seem used.

File: build.rs [hunk 3]
```diff
@@ -1,73 +1,36 @@
...
-                .arg("src/mesa/program/lex.yy.c")
-                .arg("src/compiler/glsl/glcpp/glcpp-parse.c")
-                .arg("src/compiler/glsl/glsl_parser.cpp")
-                .arg("src/mesa/program/program_parse.tab.c")
-                .arg("src/compiler/glsl/glsl_lexer.cpp")
-                .arg("src/glx/indirect.h")
-                .arg("src/glx/indirect_init.c")
-                .arg("src/mapi/glapi/glapi_mapi_tmp.h"));
+    let _dst = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let build_folder = src.join("mesa-src").join("build");
```
Comment (emilio) [inline build.rs]: A bit weird to do a build in the srcdir, but fine. Can it be in the out dir instead?

**Comment by kvark** [comment]
CI is happy now, thanks for the hint @jdm !

**Review by emilio — APPROVED** [review]
r=me with something like this applied, which seems to work just fine, if you agree...

```
diff --git a/.gitignore b/.gitignore
index e6ffe62e..04213914 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,3 @@
 Cargo.lock
 target/
-mesa-src/build
 *.pyc
diff --git a/Cargo.toml b/Cargo.toml
index 09394435..1859bf2d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,6 +8,3 @@ license = "MIT"
 description = "Mesa is a 3-D graphics library with an API which is very similar to that of OpenGL."
 homepage = "http://mesa3d.org"
 edition = "2018"
-
-[build-dependencies]
-meson = "1"
diff --git a/build.rs b/build.rs
index ae91df09..fa0604ea 100644
--- a/build.rs
+++ b/build.rs
@@ -4,14 +4,14 @@ use std::{env, fs};
 
 fn main() {
     let src = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
-    let _dst = PathBuf::from(env::var_os("OUT_DIR").unwrap());
-    let build_folder = src.join("mesa-src").join("build");
+    let dst = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let build_folder = dst.join("mesa-build");
     let _ = fs::create_dir(&build_folder);
 
     if !build_folder.join("build.ninja").exists() {
         run(Command::new("meson")
             .current_dir(&build_folder)
-            .arg("..")
+            .arg(src.join("mesa-src"))
             .arg("-Dplatforms=")
             .arg("-Ddri3=disabled")
             .arg("-Dglx-direct=false")
```

**Comment by kvark** [comment]
Thanks @emilio ! I agree with the suggestions, everything should be good now.

**Comment by kvark** [comment]
I quickly tested on WR. Clearly, the reftest expectations need to be adjusted, but it works otherwise.

**Pull request merged**